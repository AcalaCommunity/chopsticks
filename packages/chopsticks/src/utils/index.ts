export * from './decoder'
export * from './generate-html-diff'
export * from './open-html'
export * from './override'
export * from './signFake'