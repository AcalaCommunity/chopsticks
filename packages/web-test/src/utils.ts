import { ApiPromise } from '@polkadot/api'
import { ChopsticksProvider, ChopsticksProviderProps } from '@acala-network/chopsticks-core'

// for console access of chopsticks provider
export const setupChopsticksApiPromise = async (props: ChopsticksProviderProps) => {
  try {
    const provider = new ChopsticksProvider(props)
    const api = new ApiPromise({
      provider,
    })
    globalThis.api = api
    await api.isReady
    return api
  } catch (e) {
    console.log(e)
  }
}
