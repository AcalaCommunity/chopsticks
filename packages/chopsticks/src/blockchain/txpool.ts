import { EventEmitter } from 'node:stream'
import { HexString } from '@polkadot/util/types'
import _ from 'lodash'

import { Blockchain } from '.'
import { Deferred, defer } from '../utils'
import { InherentProvider } from './inherent'
import { buildBlock } from './block-builder'

export const APPLY_EXTRINSIC_ERROR = 'TxPool::ApplyExtrinsicError'

export enum BuildBlockMode {
  Batch, // one block per batch, default
  Instant, // one block per tx
  Manual, // only build when triggered
}

export interface DownwardMessage {
  sentAt: number
  msg: HexString
}

export interface HorizontalMessage {
  sentAt: number
  data: HexString
}

export interface BuildBlockParams {
  inherent?: {
    downwardMessages?: DownwardMessage[]
    horizontalMessages?: Record<number, HorizontalMessage[]>
  }
}

export class TxPool {
  readonly #chain: Blockchain
  readonly #pool: HexString[] = []
  readonly #mode: BuildBlockMode
  readonly #inherentProvider: InherentProvider
  readonly #pendingBlocks: { params: BuildBlockParams; deferred: Deferred<void> }[] = []

  readonly event = new EventEmitter()

  #isBuilding = false

  constructor(chain: Blockchain, inherentProvider: InherentProvider, mode: BuildBlockMode = BuildBlockMode.Batch) {
    this.#chain = chain
    this.#mode = mode
    this.#inherentProvider = inherentProvider
  }

  get pendingExtrinsics(): HexString[] {
    return this.#pool
  }

  submitExtrinsic(extrinsic: HexString) {
    this.#pool.push(extrinsic)

    switch (this.#mode) {
      case BuildBlockMode.Batch:
        this.#batchBuildBlock()
        break
      case BuildBlockMode.Instant:
        this.buildBlock()
        break
      case BuildBlockMode.Manual:
        // does nothing
        break
    }
  }

  #batchBuildBlock = _.debounce(this.buildBlock, 100, { maxWait: 1000 })

  async buildBlock(params?: BuildBlockParams) {
    this.#pendingBlocks.push({
      params: params || {},
      deferred: defer<void>(),
    })
    this.#buildBlockIfNeeded()
    await this.upcomingBlocks()
  }

  async upcomingBlocks() {
    const count = this.#pendingBlocks.length
    await this.#pendingBlocks[count - 1].deferred.promise
    return count
  }

  async #buildBlockIfNeeded() {
    if (this.#isBuilding) return
    if (this.#pendingBlocks.length === 0) return

    this.#isBuilding = true
    try {
      await this.#buildBlock()
    } finally {
      this.#isBuilding = false
      this.#buildBlockIfNeeded()
    }
  }

  async #buildBlock() {
    await this.#chain.api.isReady

    const pending = this.#pendingBlocks[0]
    if (!pending) {
      throw new Error('Unreachable')
    }
    const { params, deferred } = pending

    const head = this.#chain.head
    const extrinsics = this.#pool.splice(0)
    const inherents = await this.#inherentProvider.createInherents(head, params?.inherent)
    const [newBlock, pendingExtrinsics] = await buildBlock(head, inherents, extrinsics, (extrinsic, error) => {
      this.event.emit(APPLY_EXTRINSIC_ERROR, [extrinsic, error])
    })
    this.#pool.push(...pendingExtrinsics)
    await this.#chain.setHead(newBlock)

    this.#pendingBlocks.shift()
    deferred.resolve()
  }
}
