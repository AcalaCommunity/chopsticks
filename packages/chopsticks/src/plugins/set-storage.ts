import { HexString } from '@polkadot/util/types'

import { Handler, ResponseError } from '../rpc/shared'
import { StorageValues, setStorage } from '../utils/set-storage'
import { logger } from '.'

export const dev_setStorage: Handler = async (context, params) => {
  const [values, blockHash] = params as [StorageValues, HexString?]
  const hash = await setStorage(context.chain, values, blockHash).catch((error) => {
    throw new ResponseError(1, error.toString())
  })
  logger.debug(
    {
      hash,
      values,
    },
    'dev_setStorage',
  )
  return hash
}
