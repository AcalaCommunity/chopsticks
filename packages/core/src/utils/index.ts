import { HexString } from '@polkadot/util/types'
import { StorageKey } from '@polkadot/types'
import { compactStripLength, hexToU8a, u8aToHex } from '@polkadot/util'
import { hexAddPrefix, hexStripPrefix } from '@polkadot/util/hex'

import { Blockchain } from '../blockchain'

export * from './set-storage'
export * from './time-travel'
export * from './decoder'

export type GetKeys = (startKey?: string) => Promise<StorageKey<any>[]>

export type ProcessKey = (key: StorageKey<any>) => any

export async function fetchKeys(getKeys: GetKeys, processKey: ProcessKey) {
  const processKeys = async (keys: StorageKey<any>[]) => {
    for (const key of keys) {
      await processKey(key)
    }

    if (keys.length > 0) {
      return keys[keys.length - 1]
    }

    return undefined
  }

  const keys = await getKeys()
  let nextKey = await processKeys(keys)
  while (nextKey) {
    const keys = await getKeys(nextKey.toHex())
    nextKey = await processKeys(keys)
  }
}

export async function fetchKeysToArray(getKeys: GetKeys) {
  const res = [] as StorageKey<any>[]
  await fetchKeys(getKeys, (key) => res.push(key))
  return res
}

export const compactHex = (value: Uint8Array): HexString => {
  return u8aToHex(compactStripLength(value)[1])
}

export const getParaId = async (chain: Blockchain) => {
  const meta = await chain.head.meta
  const raw = await chain.head.get(compactHex(meta.query.parachainInfo.parachainId()))
  if (!raw) throw new Error('Cannot find parachain id')
  return meta.registry.createType('u32', hexToU8a(raw))
}

export const isUrl = (url: string) => {
  try {
    new URL(url)
    return true
  } catch (e) {
    return false
  }
}

export type Deferred<T> = {
  resolve: (value: T | PromiseLike<T>) => void
  reject: (reason?: any) => void
  promise: Promise<T>
}

export function defer<T>() {
  const deferred = {} as Deferred<T>
  deferred.promise = new Promise((resolve, reject) => {
    deferred.resolve = resolve
    deferred.reject = reject
  })
  return deferred
}

export const mergeKey = (...keys: (string | undefined)[]) =>
  hexAddPrefix(
    keys
      .filter((x) => !!x)
      .map((hex) => hexStripPrefix(hex))
      .join(''),
  )

const DEFAULT_CHILD_STORAGE = '0x3a6368696c645f73746f726167653a64656661756c743a'
const CHILD_LENGTH = DEFAULT_CHILD_STORAGE.length + 64

export const isChild = (key: string) => key.startsWith(DEFAULT_CHILD_STORAGE)

// returns a key that is split into the child storage key and the rest
export const splitChild = (key: string) => {
  if (!key.startsWith(DEFAULT_CHILD_STORAGE)) return
  if (key.length < CHILD_LENGTH) return
  const child = key.slice(0, CHILD_LENGTH)
  const rest = key.slice(CHILD_LENGTH)
  return [child, hexAddPrefix(rest)]
}

// returns a key that is stripped of the child storage key
export const stripChild = (key: string) => {
  const parts = splitChild(key)
  if (!parts) return key
  return parts[1]
}
