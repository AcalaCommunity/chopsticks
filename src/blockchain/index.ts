import { DataSource } from 'typeorm'
import { HexString } from '@polkadot/util/types'
import { blake2AsHex } from '@polkadot/util-crypto'
import { u8aConcat, u8aToHex } from '@polkadot/util'
import type { TransactionValidity } from '@polkadot/types/interfaces/txqueue'

import { Api } from '../api'
import { Block } from './block'
import { BuildBlockMode, BuildBlockParams, TxPool } from './txpool'
import { HeadState } from './head-state'
import { InherentProvider } from './inherent'
import { defaultLogger } from '../logger'

const logger = defaultLogger.child({ name: 'blockchain' })

export interface Options {
  api: Api
  buildBlockMode?: BuildBlockMode
  inherentProvider: InherentProvider
  db?: DataSource
  header: { number: number; hash: HexString }
  mockSignatureHost?: boolean
  allowUnresolvedImports?: boolean
}

export class Blockchain {
  readonly api: Api
  readonly db: DataSource | undefined
  readonly mockSignatureHost: boolean
  readonly allowUnresolvedImports: boolean

  readonly #txpool: TxPool

  #head: Block
  readonly #blocksByNumber: Block[] = []
  readonly #blocksByHash: Record<string, Block> = {}
  readonly #loadingBlocks: Record<string, Promise<void>> = {}

  readonly headState: HeadState

  constructor({
    api,
    buildBlockMode,
    inherentProvider,
    db,
    header,
    mockSignatureHost = false,
    allowUnresolvedImports = false,
  }: Options) {
    this.api = api
    this.db = db
    this.mockSignatureHost = mockSignatureHost
    this.allowUnresolvedImports = allowUnresolvedImports

    this.#head = new Block(this, header.number, header.hash)
    this.#registerBlock(this.#head)

    this.#txpool = new TxPool(this, inherentProvider, buildBlockMode)

    this.headState = new HeadState(this.#head)
  }

  #registerBlock(block: Block) {
    this.#blocksByNumber[block.number] = block
    this.#blocksByHash[block.hash] = block
  }

  get head(): Block {
    return this.#head
  }

  get pendingExtrinsics(): HexString[] {
    return this.#txpool.pendingExtrinsics
  }

  async getBlockAt(number?: number): Promise<Block | undefined> {
    if (number === undefined) {
      return this.head
    }
    if (number > this.#head.number) {
      return undefined
    }
    if (!this.#blocksByNumber[number]) {
      const hash = await this.api.getBlockHash(number)
      const block = new Block(this, number, hash)
      this.#registerBlock(block)
    }
    return this.#blocksByNumber[number]
  }

  async getBlock(hash?: HexString): Promise<Block | undefined> {
    await this.api.isReady
    if (hash == null) {
      hash = this.head.hash
    }
    if (!this.#blocksByHash[hash]) {
      const loadingBlock = this.#loadingBlocks[hash]
      if (loadingBlock) {
        await loadingBlock
      } else {
        const loadingBlock = (async () => {
          try {
            const header = await this.api.getHeader(hash)
            const block = new Block(this, Number(header.number), hash)
            this.#registerBlock(block)
          } catch (e) {
            logger.debug(`getBlock(${hash}) failed: ${e}`)
          }
        })()
        this.#loadingBlocks[hash] = loadingBlock
        await loadingBlock
        delete this.#loadingBlocks[hash]
      }
    }
    return this.#blocksByHash[hash]
  }

  unregisterBlock(block: Block): void {
    if (block.hash === this.head.hash) {
      throw new Error('Cannot unregister head block')
    }
    if (this.#blocksByNumber[block.number]?.hash === block.hash) {
      delete this.#blocksByNumber[block.number]
    }
    delete this.#blocksByHash[block.hash]
  }

  async setHead(block: Block): Promise<void> {
    logger.debug(
      {
        number: block.number,
        hash: block.hash,
      },
      'setHead'
    )
    this.#head = block
    this.#registerBlock(block)
    await this.headState.setHead(block)
  }

  async submitExtrinsic(extrinsic: HexString): Promise<HexString> {
    const source = '0x02' // External
    const args = u8aToHex(u8aConcat(source, extrinsic, this.head.hash))
    const res = await this.head.call('TaggedTransactionQueue_validate_transaction', args)
    const registry = await this.head.registry
    const validity: TransactionValidity = registry.createType('TransactionValidity', res.result)
    if (validity.isOk) {
      this.#txpool.submitExtrinsic(extrinsic)
      return blake2AsHex(extrinsic, 256)
    }
    throw new Error(`Extrinsic is invalid: ${validity.asErr.toString()}`)
  }

  async newBlock(params?: BuildBlockParams): Promise<Block> {
    await this.#txpool.buildBlock(params)
    return this.#head
  }
}
