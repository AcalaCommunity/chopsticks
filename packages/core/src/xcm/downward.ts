import { hexToU8a } from '@polkadot/util'

import { Blockchain } from '../blockchain/index.js'
import { DownwardMessage } from '../blockchain/txpool.js'
import { compactHex, getParaId } from '../utils/index.js'
import { setStorage } from '../utils/set-storage.js'
import { xcmLogger } from './index.js'

export const connectDownward = async (relaychain: Blockchain, parachain: Blockchain) => {
  const relayMeta = await relaychain.head.meta
  const paraId = await getParaId(parachain)
  const downwardMessageQueuesKey = compactHex(relayMeta.query.dmp.downwardMessageQueues(paraId))

  const paraMeta = await parachain.head.meta

  await relaychain.headState.subscribeStorage([downwardMessageQueuesKey], async (head, pairs) => {
    const value = pairs[0][1]
    if (!value) return

    const meta = await head.meta
    const downwardMessageQueuesKey = compactHex(meta.query.dmp.downwardMessageQueues(paraId))

    // clear relaychain message queue
    await setStorage(relaychain, [[downwardMessageQueuesKey, null]], head.hash)

    const downwardMessages = meta.registry
      .createType('Vec<PolkadotCorePrimitivesInboundDownwardMessage>', hexToU8a(value))
      .toJSON() as any as DownwardMessage[]

    if (downwardMessages.length === 0) return

    xcmLogger.debug({ downwardMessages }, 'downward_message')
    parachain.submitDownwardMessages(downwardMessages)

    // We need to produce an extra block to process the message from the queue
    // in case `MessageQueue` is used for dmp.
    if (paraMeta.tx.messageQueue) {
      await parachain.newBlock()
    }
  })
}
