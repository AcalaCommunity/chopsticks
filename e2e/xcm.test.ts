import { afterAll, describe, it } from 'vitest'

import { DownwardMessage, HorizontalMessage } from '../src/blockchain/txpool'
import { connectDownward } from '../src/xcm/downward'
import { matchSnapshot, setupAll, testingPairs } from './helper'
import { setStorage } from '../src/utils/set-storage'

const downwardMessages: DownwardMessage[] = [
  {
    sentAt: 1,
    msg: '0x0210010400010000078155a74e390a1300010000078155a74e39010300286bee0d01000400010100c0cbffafddbe39f71f0190c2369adfc59eaa4c81a308ebcad88cdd9c400ba57c',
  },
]

const horizontalMessages: Record<number, HorizontalMessage[]> = {
  2004: [
    {
      data: '0x000210000400000106080001000fc2ddd331d55e200a1300000106080001000fc2ddd331d55e20010700f2052a010d01000400010100ba686c8fa59178c699a698ea4d8e2c595394c2594bce4b6c2ca3a9bf3018e25d',
      sentAt: 13509121,
    },
  ],
}

describe('XCM', async () => {
  const ctxAcala = await setupAll({
    endpoint: 'wss://acala-rpc-1.aca-api.network',
    blockHash: '0x663c25dc86521f4b7f74dcbc26224bb0fac40e316e6b0bcf6a51de373f37afac',
  })

  const ctxPolkadot = await setupAll({
    endpoint: 'wss://rpc.polkadot.io',
    blockHash: '0x0a26b277b252fc61efcda02e44e95c73bf7ae21233bacb2d3bd7631212350d59',
  })

  afterAll(async () => {
    await ctxAcala.teardownAll()
    await ctxPolkadot.teardownAll()
  })

  it('Acala handles downward messages', async () => {
    const { chain, api, teardown } = await ctxAcala.setup()
    await chain.newBlock({ inherent: { downwardMessages } })
    await matchSnapshot(api.query.system.events())
    await teardown()
  })

  it('Acala handles horizonal messages', async () => {
    const { chain, api, teardown } = await ctxAcala.setup()
    await chain.newBlock({ inherent: { horizontalMessages } })
    await matchSnapshot(api.query.system.events())
    await teardown()
  })

  it('Polkadot send downward messages to Acala', async () => {
    const polkadot = await ctxPolkadot.setup()
    const acala = await ctxAcala.setup()

    await connectDownward(polkadot.chain, acala.chain)

    const { alice } = testingPairs()

    await setStorage(polkadot.chain, {
      System: {
        Account: [[[alice.address], { data: { free: 1000 * 1e10 } }]],
      },
    })

    await polkadot.api.tx.xcmPallet
      .reserveTransferAssets(
        { V0: { X1: { Parachain: 2000 } } },
        {
          V0: {
            X1: {
              AccountId32: {
                network: 'Any',
                id: alice.addressRaw,
              },
            },
          },
        },
        {
          V0: [
            {
              ConcreteFungible: { id: 'Null', amount: 100e10 },
            },
          ],
        },
        0
      )
      .signAndSend(alice)

    await polkadot.chain.newBlock()
    await matchSnapshot(polkadot.api.query.system.events())
    await matchSnapshot(acala.api.query.system.events())

    await polkadot.teardown()
    await acala.teardown()
  })
})
