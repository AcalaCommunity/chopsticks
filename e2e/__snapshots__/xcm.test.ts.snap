// Vitest Snapshot v1

exports[`XCM > Acala handles downward messages 1`] = `
[
  {
    "event": {
      "data": {
        "dispatchInfo": {
          "class": "Mandatory",
          "paysFee": "Yes",
          "weight": "158,080,000",
        },
      },
      "index": "0x0000",
      "method": "ExtrinsicSuccess",
      "section": "system",
    },
    "phase": {
      "ApplyExtrinsic": "0",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "count": "1",
      },
      "index": "0x1e04",
      "method": "DownwardMessagesReceived",
      "section": "parachainSystem",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "amount": "246,129,953,122",
        "currencyId": {
          "Token": "DOT",
        },
        "who": "25EPNyxpdqMUeVj9WndwUYVKkfFkMLroo2fg9V1MA79PJ6iE",
      },
      "index": "0x0b0a",
      "method": "Deposited",
      "section": "tokens",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "account": "25EPNyxpdqMUeVj9WndwUYVKkfFkMLroo2fg9V1MA79PJ6iE",
      },
      "index": "0x0003",
      "method": "NewAccount",
      "section": "system",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "amount": "246,129,953,122",
        "currencyId": {
          "Token": "DOT",
        },
        "who": "25EPNyxpdqMUeVj9WndwUYVKkfFkMLroo2fg9V1MA79PJ6iE",
      },
      "index": "0x0b00",
      "method": "Endowed",
      "section": "tokens",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "amount": "2,771,999",
        "currencyId": {
          "Token": "DOT",
        },
        "who": "23M5ttkmR6KcoTAAE6gcmibnKFtVaTP5yxnY8HF1BmrJ2A1i",
      },
      "index": "0x0b0a",
      "method": "Deposited",
      "section": "tokens",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "messageId": "0xa004205cd3559f06413e92aa4bd5a0f4a12eea722da2bf27c52447f69fa7b660",
        "outcome": {
          "Complete": "800,000,000",
        },
      },
      "index": "0x3502",
      "method": "ExecutedDownward",
      "section": "dmpQueue",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "dmqHead": "0x43bc1b7d0a62cc18a3ac5218f8076cfce521669761c7cc68761a4094167ab974",
        "weightUsed": "800,000,000",
      },
      "index": "0x1e05",
      "method": "DownwardMessagesProcessed",
      "section": "parachainSystem",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "dispatchInfo": {
          "class": "Mandatory",
          "paysFee": "Yes",
          "weight": "0",
        },
      },
      "index": "0x0000",
      "method": "ExtrinsicSuccess",
      "section": "system",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
]
`;

exports[`XCM > Acala handles horizonal messages 1`] = `
[
  {
    "event": {
      "data": {
        "dispatchInfo": {
          "class": "Mandatory",
          "paysFee": "Yes",
          "weight": "158,080,000",
        },
      },
      "index": "0x0000",
      "method": "ExtrinsicSuccess",
      "section": "system",
    },
    "phase": {
      "ApplyExtrinsic": "0",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "amount": "9,111,469,011,754,434",
        "currencyId": {
          "Token": "AUSD",
        },
        "who": "23UvQ3ZQXJ5LfTUSYkcRPkQX2FHgcKxGmqdxYJe9j5e3Lwsi",
      },
      "index": "0x0b08",
      "method": "Withdrawn",
      "section": "tokens",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "amount": "9,111,466,917,717,450",
        "currencyId": {
          "Token": "AUSD",
        },
        "who": "2561WgnRM2Nbec2WseHxVdpeW19YYHh7RM5gRWQV7zwwwgXQ",
      },
      "index": "0x0b0a",
      "method": "Deposited",
      "section": "tokens",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "account": "2561WgnRM2Nbec2WseHxVdpeW19YYHh7RM5gRWQV7zwwwgXQ",
      },
      "index": "0x0003",
      "method": "NewAccount",
      "section": "system",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "amount": "9,111,466,917,717,450",
        "currencyId": {
          "Token": "AUSD",
        },
        "who": "2561WgnRM2Nbec2WseHxVdpeW19YYHh7RM5gRWQV7zwwwgXQ",
      },
      "index": "0x0b00",
      "method": "Endowed",
      "section": "tokens",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "amount": "2,094,036,984",
        "currencyId": {
          "Token": "AUSD",
        },
        "who": "23M5ttkmR6KcoTAAE6gcmibnKFtVaTP5yxnY8HF1BmrJ2A1i",
      },
      "index": "0x0b0a",
      "method": "Deposited",
      "section": "tokens",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "messageHash": "0x09e1b5b4cd5b9ab2cb31449175dc5504e4314db308d38de8dfd8d6529e481b79",
        "weight": "800,000,000",
      },
      "index": "0x3200",
      "method": "Success",
      "section": "xcmpQueue",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "dispatchInfo": {
          "class": "Mandatory",
          "paysFee": "Yes",
          "weight": "0",
        },
      },
      "index": "0x0000",
      "method": "ExtrinsicSuccess",
      "section": "system",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
]
`;

exports[`XCM > Polkadot send downward messages to Acala 1`] = `
[
  {
    "event": {
      "data": {
        "dispatchInfo": {
          "class": "Mandatory",
          "paysFee": "Yes",
          "weight": "218,918,000",
        },
      },
      "index": "0x0000",
      "method": "ExtrinsicSuccess",
      "section": "system",
    },
    "phase": {
      "ApplyExtrinsic": "0",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "dispatchInfo": {
          "class": "Mandatory",
          "paysFee": "Yes",
          "weight": "85,212,000",
        },
      },
      "index": "0x0000",
      "method": "ExtrinsicSuccess",
      "section": "system",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "amount": "184,908,979",
        "who": "146SvjUZXoMaemdeiecyxgALeYMm8ZWh1yrGo8RtpoPfe7WL",
      },
      "index": "0x0508",
      "method": "Withdraw",
      "section": "balances",
    },
    "phase": {
      "ApplyExtrinsic": "2",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "amount": "1,000,000,000,000",
        "from": "146SvjUZXoMaemdeiecyxgALeYMm8ZWh1yrGo8RtpoPfe7WL",
        "to": "13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy",
      },
      "index": "0x0502",
      "method": "Transfer",
      "section": "balances",
    },
    "phase": {
      "ApplyExtrinsic": "2",
    },
    "topics": [],
  },
  {
    "event": {
      "data": [
        {
          "Complete": "1,000,000,000",
        },
      ],
      "index": "0x6300",
      "method": "Attempted",
      "section": "xcmPallet",
    },
    "phase": {
      "ApplyExtrinsic": "2",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "amount": "147,927,183",
        "who": "13UVJyLnbVp9RBZYFwFGyDvVd1y27Tt8tkntv6Q7JVPhFsTB",
      },
      "index": "0x0507",
      "method": "Deposit",
      "section": "balances",
    },
    "phase": {
      "ApplyExtrinsic": "2",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "value": "147,927,183",
      },
      "index": "0x1306",
      "method": "Deposit",
      "section": "treasury",
    },
    "phase": {
      "ApplyExtrinsic": "2",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "amount": "36,981,796",
        "who": "12HFymxpDmi4XXPHaEMp74CNpRhkqwG5qxnrgikkhon1XMrj",
      },
      "index": "0x0507",
      "method": "Deposit",
      "section": "balances",
    },
    "phase": {
      "ApplyExtrinsic": "2",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "actualFee": "184,908,979",
        "tip": "0",
        "who": "146SvjUZXoMaemdeiecyxgALeYMm8ZWh1yrGo8RtpoPfe7WL",
      },
      "index": "0x2000",
      "method": "TransactionFeePaid",
      "section": "transactionPayment",
    },
    "phase": {
      "ApplyExtrinsic": "2",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "dispatchInfo": {
          "class": "Normal",
          "paysFee": "Yes",
          "weight": "1,185,212,000",
        },
      },
      "index": "0x0000",
      "method": "ExtrinsicSuccess",
      "section": "system",
    },
    "phase": {
      "ApplyExtrinsic": "2",
    },
    "topics": [],
  },
]
`;

exports[`XCM > Polkadot send downward messages to Acala 2`] = `
[
  {
    "event": {
      "data": {
        "dispatchInfo": {
          "class": "Mandatory",
          "paysFee": "Yes",
          "weight": "158,080,000",
        },
      },
      "index": "0x0000",
      "method": "ExtrinsicSuccess",
      "section": "system",
    },
    "phase": {
      "ApplyExtrinsic": "0",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "count": "2",
      },
      "index": "0x1e04",
      "method": "DownwardMessagesReceived",
      "section": "parachainSystem",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "amount": "235,975,337,226",
        "currencyId": {
          "Token": "DOT",
        },
        "who": "23e3iqWXzeCN5H3aJzDmqk2bz1mZ9pRG94PM1NJdfBaMj6ZR",
      },
      "index": "0x0b0a",
      "method": "Deposited",
      "section": "tokens",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "amount": "235,975,337,226",
        "currencyId": {
          "Token": "DOT",
        },
        "who": "23e3iqWXzeCN5H3aJzDmqk2bz1mZ9pRG94PM1NJdfBaMj6ZR",
      },
      "index": "0x0b00",
      "method": "Endowed",
      "section": "tokens",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "amount": "2,771,999",
        "currencyId": {
          "Token": "DOT",
        },
        "who": "23M5ttkmR6KcoTAAE6gcmibnKFtVaTP5yxnY8HF1BmrJ2A1i",
      },
      "index": "0x0b0a",
      "method": "Deposited",
      "section": "tokens",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "messageId": "0x8088b1e1a4d66511b52e8a8f6155e42241ea1818151625a924852c814d9aa265",
        "outcome": {
          "Complete": "800,000,000",
        },
      },
      "index": "0x3502",
      "method": "ExecutedDownward",
      "section": "dmpQueue",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "amount": "999,997,228,001",
        "currencyId": {
          "Token": "DOT",
        },
        "who": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
      },
      "index": "0x0b0a",
      "method": "Deposited",
      "section": "tokens",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "account": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
      },
      "index": "0x0003",
      "method": "NewAccount",
      "section": "system",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "amount": "999,997,228,001",
        "currencyId": {
          "Token": "DOT",
        },
        "who": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
      },
      "index": "0x0b00",
      "method": "Endowed",
      "section": "tokens",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "amount": "2,771,999",
        "currencyId": {
          "Token": "DOT",
        },
        "who": "23M5ttkmR6KcoTAAE6gcmibnKFtVaTP5yxnY8HF1BmrJ2A1i",
      },
      "index": "0x0b0a",
      "method": "Deposited",
      "section": "tokens",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "messageId": "0xc4dd612e0a23c9e0d3c677adbfb3bca299caa4c37c2085baa5d7a33736d19e80",
        "outcome": {
          "Complete": "800,000,000",
        },
      },
      "index": "0x3502",
      "method": "ExecutedDownward",
      "section": "dmpQueue",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "dmqHead": "0xd6308f1fd114709106d34c2cecce1bdd2fd9fec73898e25ccd5c62f84a96f8bd",
        "weightUsed": "1,600,000,000",
      },
      "index": "0x1e05",
      "method": "DownwardMessagesProcessed",
      "section": "parachainSystem",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
  {
    "event": {
      "data": {
        "dispatchInfo": {
          "class": "Mandatory",
          "paysFee": "Yes",
          "weight": "0",
        },
      },
      "index": "0x0000",
      "method": "ExtrinsicSuccess",
      "section": "system",
    },
    "phase": {
      "ApplyExtrinsic": "1",
    },
    "topics": [],
  },
]
`;
